subprojects/core/src/main/java/org/gradle/api/internal/collections/IterationOrderRetainingSetElementSource.java

```diff
@@ -46,6 +46,10 @@ public boolean isSatisfiedBy(ValuePointer<T> pointer) {
 
     @Override
     public boolean add(T element) {
+        if (isRealizing()) {
+            return true;
```

Comment:
🔍 This tries to fix the difference between `ElementSource` implementation where an element been resolved would kick an add inside the container which would succeed and duplicate the element. This should be reviewed deeply to ensure we aren't breaking something. Don't hesitate to comment on test coverage that we could add as well.